pkg/scheduler/framework/plugins/topologyspreadconstraints/utils_test.go

```diff
@@ -12,19 +12,512 @@ import (
+				smallest:       -1,
+				secondSmallest: -1,
+				largest:        -1,
+			},
+		},
+		{
+			name: "uninitialized (no topology key present in clusters)",
+			clusters: []fleetv1beta1.MemberCluster{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: clusterName1,
+					},
+				},
+			},
+			bindings: []*fleetv1beta1.ClusterResourceBinding{},
+			wantBindingCounterByDomain: &bindingCounterByDomain{
+				counter:        make(map[domainName]count),
+				smallest:       -1,
+				secondSmallest: -1,
+				largest:        -1,
+			},
+		},
+		{
+			name: "single domain",
+			clusters: []fleetv1beta1.MemberCluster{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: clusterName1,
+						Labels: map[string]string{
+							topologyKey1: topologyValue1,
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: clusterName2,
+						Labels: map[string]string{
+							topologyKey1: topologyValue1,
+						},
+					},
+				},
+			},
+			bindings: []*fleetv1beta1.ClusterResourceBinding{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: bindingName1,
+					},
+					Spec: fleetv1beta1.ResourceBindingSpec{
+						TargetCluster: clusterName1,
+					},
+				},
+			},
+			wantBindingCounterByDomain: &bindingCounterByDomain{
+				counter: map[domainName]count{
+					topologyValue1: 1,
+				},
+				smallest:       1,
+				secondSmallest: 1,
+				largest:        1,
+			},
+		},
+		{
+			name: "multiple domains, same count",
+			clusters: []fleetv1beta1.MemberCluster{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: clusterName1,
+						Labels: map[string]string{
+							topologyKey1: topologyValue1,
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: clusterName2,
+						Labels: map[string]string{
+							topologyKey1: topologyValue2,
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: clusterName3,
+						Labels: map[string]string{
+							topologyKey1: topologyValue3,
+						},
+					},
+				},
+			},
+			bindings: []*fleetv1beta1.ClusterResourceBinding{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: bindingName1,
+					},
+					Spec: fleetv1beta1.ResourceBindingSpec{
+						TargetCluster: clusterName1,
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: bindingName2,
+					},
+					Spec: fleetv1beta1.ResourceBindingSpec{
+						TargetCluster: clusterName2,
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: bindingName3,
+					},
+					Spec: fleetv1beta1.ResourceBindingSpec{
+						TargetCluster: clusterName3,
+					},
+				},
+			},
+			wantBindingCounterByDomain: &bindingCounterByDomain{
+				counter: map[domainName]count{
+					topologyValue1: 1,
+					topologyValue2: 1,
+					topologyValue3: 1,
+				},
+				smallest:       1,
+				secondSmallest: 1,
+				largest:        1,
+			},
+		},
+		{
+			name: "multiple domains, different counts, separate special counts",
```

Comment:
hmm, i did not get what's the "separate special" counts. could you please explain it a little bit?

Reply:
The separate special counts refer to the smallest, second smallest, and largest counts the counter tracks.